ssl/record/rec_layer_s3.c

```diff
@@ -751,10 +751,29 @@ int do_ssl3_write(SSL *s, int type, const unsigned char *buf,
             size_t tmppipelen = 0;
             int ret;
 
-            ret = do_ssl3_write(s, type, buf, &tmppipelen, 1, 1, &prefix_len);
-            if (ret <= 0) {
-                /* SSLfatal() already called if appropriate */
-                goto err;
+            if (BIO_get_ktls_send(s->wbio)) {
+                /*
+                 * For KTLS, send an explicit empty fragment rather than
+                 * storing the fragment as a prefix in the buffer.
+                 */
+                ret = BIO_flush(s->wbio);
```

Comment:
Why is this flush necessary?

Reply:
Anytime with KTLS that we wish to send an frame with an explicit message, we must flush any pending data (that gets written as application data) before calling sendmsg() to send out a frame with a specific type and length.

Reply:
Ok. But please set `s->rwstate = SSL_WRITING;` before this call.

Reply:
Should I set it back to `SSL_NOTHING` if `BIO_flush` fails?

Reply:
If its a retryable failure then we leave it as `SSL_WRITING` which indicates to `SSL_get_error()` that we should retry.